src/vptranslator.rs

```diff
@@ -0,0 +1,253 @@
+#[derive(Debug)]
+pub struct ExceptionDate {
+    ///Date of exception
+    pub date: Date,
+    ///exception type
+    pub exception_type: ExceptionType,
+}
+
+impl PartialEq for ExceptionDate {
+    fn eq(&self, other: &ExceptionDate) -> bool {
+        self.exception_type == other.exception_type
+    }
+}
+
+///Presents a list of dates in the form of intervals and exception dates.
+#[derive(Default, Debug)]
+pub struct BlockPattern {
+    ///Indicates operating days of the service
+    pub operating_days: Vec<Weekday>,
+    ///Start and end service day for the service interval
+    pub validity_periods: Vec<ValidityPeriod>,
+    ///List of dates where service is available or not in interval.
+    pub exceptions: Vec<ExceptionDate>,
+}
+
+fn get_prev_monday(date: Date) -> Date {
+    date + Duration::days(-1 * date.weekday().num_days_from_monday() as i64)
+}
+
+fn compute_validity_pattern(start_date: Date, end_date: Date, dates: &BTreeSet<Date>) -> Vec<u8> {
+    let length = (end_date.signed_duration_since(start_date).num_weeks() + 1) as usize;
+    let mut res = vec![0; length];
+    for date in dates {
+        let w = date.signed_duration_since(start_date).num_weeks() as usize;
+        res[w] |= 1 << (7 - date.weekday().number_from_monday());
+    }
+    res
+}
+
+fn dists(w: u8, weeks: &[u8]) -> u32 {
+    let res = weeks
+        .iter()
+        .filter(|n| **n != 0u8)
+        .map(|n| (*n as u8 ^ w).count_ones())
+        .sum();
+    res
+}
+
+fn get_min_week_pattern(weeks: &[u8]) -> u8 {
+    let mut best: u8 = 0;
+    let mut best_score = std::u32::MAX;
+    for i in 0..128 {
+        let score = dists(i, weeks);
+        if (score < best_score)
+            || ((score == best_score) && (score.count_ones() < best_score.count_ones()))
+        {
+            best_score = score;
+            best = i;
+        }
+    }
+    best
+}
+
+fn fill_exceptions(
+    start_date: Date,
+    exception: u8,
+    exception_type: ExceptionType,
+    exception_list: &mut Vec<ExceptionDate>,
+) {
+    for i in 0..7 {
+        if exception & (1 << i) == (1 << i) {
+            let date = start_date + Duration::days((6 - i) as i64);
+            exception_list.push(ExceptionDate {
+                date: date,
+                exception_type: exception_type.clone(),
+            });
+        }
+    }
+}
+
+fn fill_operating_days(week: u8, operating_days: &mut Vec<Weekday>) {
+    for i in 0..7 {
+        if week & (1 << i) == (1 << i) {
+            operating_days.push(Weekday::from_u8(6 - i).unwrap());
+        }
+    }
+    operating_days.sort_by_key(|w| w.num_days_from_monday());
+}
+
+fn clean_extra_dates(start_date: Date, end_date: Date, dates: &mut Vec<ExceptionDate>) {
+    dates.retain(|d| d.date >= start_date && d.date <= end_date);
+}
+///Allows you to present a list of dates in a readable way.
+pub fn translate(dates: &BTreeSet<Date>) -> BlockPattern {
+    let mut res = BlockPattern::default();
```

Comment:
can't you make this variable not mutable, and create it from all the fields at the end ?